### PR TITLE
openclaw,tool: guard blocked web research routes

### PR DESCRIPTION
**Pull request by WineChord**
## Summary
- block search-result and challenge pages in browser/web-fetch flows by default, with explicit opt-outs
- reject ad hoc proxy/tunnel routing attempts in chat exec commands
- add an opt-in host exec idle-wait cap for benchmark/runtime runs that should not wait out external rate limits

## Validation
- cd openclaw && go test ./internal/octool ./app
- cd openclaw && go test ./...

### CODE REVIEW

**Comment by WineChord** [comment]
fixed: added DuckDuckGo main search and Yahoo search positive coverage, and expanded blocked browser-page handling with focused tests.

**Comment by WineChord** [comment]
fixed: added focused coverage for human-verification and bot-check blocked-page detection, plus a nearby negative case.

File: tool/webfetch/httpfetch/fetch.go [hunk 1]
```diff
@@ -389,6 +448,225 @@ func unsupportedContentTypeError(contentType string) string {
...
+		Name:  "Brave Search",
+		Host:  "search.brave.com",
+		Paths: []string{"/search"},
+	},
+	{
+		Name:  "Yahoo search",
+		Host:  "search.yahoo.com",
+		Paths: []string{"/search"},
+	},
+}
```
[the remark this answers is not in the document]
Comment (WineChord) [reply tool/webfetch/httpfetch/fetch.go:522]: fixed: added query-key requirements for Google, Google Scholar, Bing, Brave, and Yahoo search-result rules in both web_fetch and browser navigation, with negative coverage for non-search /search subpages.

File: openclaw/app/blocked_route_tool_callback.go [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+	host := strings.ToLower(parsed.Hostname())
+	host = strings.TrimPrefix(host, "www.")
+	if host == "" {
+		return "", false
+	}
+	return host, true
+}
+
+func blockedRouteReason(item blockedRouteResultItem) (string, bool) {
+	evidence := strings.ToLower(item.Error + "\n" + item.Content)
```
Comment (Rememorio) [inline openclaw/app/blocked_route_tool_callback.go]: **P1: Do not infer blocked hosts from successful page content**

The new callback scans `item.Content` as well as the structured error. A normal 200 response that discusses “rate limits”, Cloudflare, CAPTCHA, or bot checks therefore marks the entire host as blocked, and later URLs on that host are skipped.

This also defeats the documented `detect_blocked_pages: false` opt-out: challenge content may be returned once, but the callback still suppresses subsequent requests. Record 429 from `StatusCode` and anti-bot evidence from the structured `Error` only, then add a regression case with successful explanatory content and detection disabled.

<details><summary>中文</summary>

**不要根据成功页面内容推断主机已被封禁**

新回调同时扫描 `item.Content` 和结构化错误。因此，一个正常的 200 页面只要讨论了“rate limits”、Cloudflare、CAPTCHA 或 bot check，就会把整个主机记为已封禁，后续访问该主机的其他 URL 都会被跳过。

这也破坏了文档中的 `detect_blocked_pages: false` 显式关闭语义：挑战页内容第一次可以返回，但回调仍会阻止后续请求。建议仅通过 `StatusCode` 判断 429，并只从结构化 `Error` 识别反爬挑战，同时增加“成功说明性内容 + 关闭检测”的回归测试。

</details>

File: openclaw/internal/octool/policy.go [hunk 3]
```diff
@@ -118,6 +187,408 @@ func blocksSystemPackageInstall(command string) bool {
...
+		strings.Trim(strings.TrimSpace(value), shellQuoteChars),
+	)
+	return ok
+}
+
+func blocksNetworkProxyDepth(command string, depth int) bool {
+	if command == "" || depth > 2 {
+		return false
+	}
+	if blocksProgrammaticNetworkProxy(command) {
```
Comment (Rememorio) [inline openclaw/internal/octool/policy.go:386]: **P1: Inspect programmatic proxies in every shell segment**

`blocksProgrammaticNetworkProxy` runs against the whole command, while its executable-position check rejects Python or Node when an earlier segment contains another command. For example, `cd /tmp && python3 -c "requests.get(..., proxies=...)"` passes this policy; the later per-segment loop only checks curl/wget/SSH flags.

Run the programmatic proxy check for each `shellPolicySegments` entry, or make executable-position analysis segment-aware. Add chained Python and Node cases to the blocking tests.

<details><summary>中文</summary>

**应检查每个 Shell 片段中的程序化代理**

`blocksProgrammaticNetworkProxy` 只检查整条命令，而它的可执行位置判断在前一个片段出现其他命令时会忽略后面的 Python 或 Node。例如，`cd /tmp && python3 -c "requests.get(..., proxies=...)"` 可以通过策略；后续按片段检查的逻辑又只覆盖 curl、wget 和 SSH 参数。

建议对 `shellPolicySegments` 的每个片段执行程序化代理检测，或让可执行位置判断感知 Shell 片段边界，并补充 Python、Node 链式命令的拦截测试。

</details>

File: tool/webfetch/httpfetch/fetch.go [hunk 4]
```diff
@@ -278,6 +319,12 @@ func (t *webFetchTool) fetchOne(ctx context.Context, urlStr string) resultItem {
 		item.Error = err.Error()
```
Comment (Rememorio) [inline tool/webfetch/httpfetch/fetch.go:319]: **P1: Apply search-page blocking to redirects**

Search-page validation only examines the submitted URL before `http.Client.Do`. Because the client follows redirects, an allowed URL or shortener can redirect to a Google/Bing/DDG result page and the response is processed normally, bypassing the new default.

Validate each redirect target while preserving any caller-supplied `CheckRedirect`, or reject the final `resp.Request.URL` before consuming its body. A local redirect-to-SERP test would close this gap.

<details><summary>中文</summary>

**搜索结果页拦截需要覆盖重定向**

当前只在调用 `http.Client.Do` 前检查用户提交的 URL。由于客户端默认跟随重定向，一个普通 URL 或短链接可以跳转到 Google、Bing 或 DuckDuckGo 搜索结果页，随后响应仍会被正常处理，从而绕过新增的默认限制。

建议在保留调用方 `CheckRedirect` 逻辑的前提下校验每个重定向目标，或至少在读取响应体前检查最终的 `resp.Request.URL`。可以用本地服务器构造跳转到搜索结果页的测试覆盖该场景。

</details>

File: openclaw/internal/octool/policy.go [hunk 5]
```diff
@@ -118,6 +187,408 @@ func blocksSystemPackageInstall(command string) bool {
...
+}
+
+func sleepDuration(words []string) (time.Duration, bool) {
+	for _, word := range words {
+		value := strings.Trim(strings.TrimSpace(word), shellQuoteChars)
+		value = strings.TrimRight(value, ".,;:)]}")
+		if value == "" || strings.HasPrefix(value, "-") {
+			continue
+		}
+		return parseIdleWaitDuration(value)
```
Comment (Rememorio) [inline openclaw/internal/octool/policy.go]: **P2: Account for all valid sleep operands**

The idle-wait parser returns after the first duration operand. GNU `sleep` accepts multiple operands and sums them, so `sleep 15 15` passes a 20-second cap but waits 30 seconds. `time.ParseDuration` also rejects the valid GNU suffix `d`, allowing commands such as `sleep 1d` through any configured cap.

Parse and sum all duration operands with overflow protection, including the supported day suffix, and add boundary tests proving the total wait cannot exceed `MaxIdleWait`.

<details><summary>中文</summary>

**需要覆盖所有合法的 sleep 参数**

空闲等待解析器在读到第一个时长参数后立即返回。GNU `sleep` 支持多个参数并将其相加，因此在上限为 20 秒时，`sleep 15 15` 会通过检查但实际等待 30 秒。`time.ParseDuration` 也不支持 GNU 合法的 `d` 后缀，导致 `sleep 1d` 可以绕过任意配置上限。

建议在防溢出的前提下解析并累加全部时长参数，同时支持 day 后缀，并增加总等待时间不得超过 `MaxIdleWait` 的边界测试。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 4 inline comments below. The comments focus on issues worth addressing before merge.

Reviewed the new OpenClaw route guards, host-level blocked-route memory, exec policy, and idle-wait cap. Four concrete gaps remain: redirects bypass search-page blocking, successful content can poison host memory despite the detection opt-out, chained commands bypass programmatic proxy detection, and valid sleep forms bypass the configured cap. Targeted regression tests for these cases would provide the most useful validation.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 4 条行内评论。评论聚焦在合并前值得处理的问题。

已审查 OpenClaw 新增的路由防护、主机级封禁记忆、exec 策略和空闲等待上限。仍有四个明确问题：重定向可绕过搜索结果页拦截，成功响应内容可能污染主机封禁状态且无视检测开关，链式命令可绕过程序化代理检测，部分合法 sleep 写法可绕过等待上限。建议优先为这些场景补充针对性回归测试。

</details>

File: openclaw/app/blocked_route_tool_callback.go (hunk 2 above)
Comment (WineChord) [reply openclaw/app/blocked_route_tool_callback.go]: fixed

File: openclaw/internal/octool/policy.go (hunk 3 above)
Comment (WineChord) [reply openclaw/internal/octool/policy.go:386]: fixed

File: tool/webfetch/httpfetch/fetch.go (hunk 4 above)
Comment (WineChord) [reply tool/webfetch/httpfetch/fetch.go:319]: fixed

File: openclaw/internal/octool/policy.go (hunk 5 above)
Comment (WineChord) [reply openclaw/internal/octool/policy.go]: fixed

File: openclaw/app/blocked_route_tool_callback.go (hunk 2 above)
Comment (Rememorio) [reply openclaw/app/blocked_route_tool_callback.go]: Confirmed fixed: successful page content is no longer used as blocked-route evidence, and the new regression test covers explanatory 200 content without suppressing the host’s next request.

<details><summary>中文</summary>

已确认修复：成功页面内容不再作为路由封禁依据，新增回归测试也覆盖了说明性 200 内容不会阻止该主机的后续请求。

</details>

File: openclaw/internal/octool/policy.go (hunk 3 above)
Comment (Rememorio) [reply openclaw/internal/octool/policy.go:386]: Confirmed fixed: programmatic proxy detection now runs per shell segment, so chained Python and Node commands are both blocked.

<details><summary>中文</summary>

已确认修复：程序化代理检测现在会逐个 Shell 片段执行，因此链式 Python 和 Node 命令都会被拦截。

</details>

File: tool/webfetch/httpfetch/fetch.go (hunk 4 above)
Comment (Rememorio) [reply tool/webfetch/httpfetch/fetch.go:319]: Confirmed fixed: the final redirect URL is validated before the body is consumed, and the regression test covers the redirect bypass while preserving the custom CheckRedirect behavior.

<details><summary>中文</summary>

确认已修复：最终重定向地址会在读取响应体前被校验，回归测试也覆盖了该绕过场景，并保留了自定义 CheckRedirect 的行为。

</details>

File: openclaw/internal/octool/policy.go (hunk 5 above)
Comment (Rememorio) [reply openclaw/internal/octool/policy.go]: Confirmed fixed. All sleep operands are now summed with overflow protection, the `d` suffix is supported, and the boundary cases are covered by tests.

<details><summary>中文</summary>

已确认修复。现在会累加所有 sleep 参数并防止溢出，同时支持 `d` 后缀，相关边界情况也已有测试覆盖。

</details>

File: openclaw/internal/browser/tool.go [hunk 6]
```diff
@@ -1581,17 +1604,25 @@ func (t *Tool) handleNavigate(
...
-	}
-
 	rawURL := browserURL(in.URL, in.TargetURL)
 	if rawURL == "" {
 		return Result{}, errors.New("browser navigate requires url")
 	}
 	if err := t.navigation.Validate(rawURL); err != nil {
 		return Result{}, err
 	}
+	if result, ok := t.searchResultBlockedResult(
```
Comment (Rememorio) [inline openclaw/internal/browser/tool.go:1631]: **P1: Enforce the search-page policy after browser navigation**

The new check validates only the requested URL before calling the driver. An allowed URL can redirect to a search-result page, and `act` operations such as clicks or form submissions can navigate there without supplying a URL to this check. The final page is then returned normally because `textResult` only detects anti-bot text.

Validate the resulting/current page URL after navigation-capable operations, or enforce the policy in the driver through navigation interception. Add redirect and click/form regression cases proving that the default policy cannot reach a SERP unless `allow_search_result_pages` is enabled.

<details><summary>中文</summary>

**在浏览器导航完成后继续执行搜索页策略**

新增检查只在调用驱动前验证请求中的 URL。一个允许访问的地址可以重定向到搜索结果页，点击或提交表单等 `act` 操作也能在不经过这段 URL 检查的情况下完成导航。最终页面仍会正常返回，因为 `textResult` 只识别反自动化页面文本。

建议在所有可能触发导航的操作完成后校验当前页面 URL，或在驱动层拦截导航请求。应增加重定向和点击/表单场景的测试，证明默认配置下无法进入搜索结果页，除非明确启用 `allow_search_result_pages`。

</details>

File: openclaw/internal/octool/policy.go [hunk 7]
```diff
@@ -91,6 +141,25 @@ func NewChatCommandSafetyPolicy() CommandPolicy {
 				reasonSystemPackageInstall,
 			)
 		}
+		if source, ok := blocksSearchResultHTTP(req.Command); ok {
+			return fmt.Errorf(
+				errCommandPolicyRejected,
+				fmt.Sprintf(reasonSearchResultHTTP, source),
+			)
+		}
+		if blocksNetworkProxy(req.Command) {
```
Comment (Rememorio) [inline openclaw/internal/octool/policy.go]: **P1: Reject proxy overrides supplied through exec env arguments**

The proxy policy inspects only `req.Command`, although `exec_command` exposes an `env` map. A call with `env: {"HTTPS_PROXY":"http://proxy:8080"}` and an otherwise ordinary `curl` command therefore routes through an ad hoc proxy without reaching `blocksNetworkProxy`.

Because the host implementation later merges these explicit values with the configured environment, inspect the tool-provided overrides before that merge or carry their provenance in `CommandRequest`. Preserve configured baseline proxies, but reject non-empty `HTTP_PROXY`, `HTTPS_PROXY`, and `ALL_PROXY` overrides, with coverage for both host and sandbox exec tools.

<details><summary>中文</summary>

**拦截通过 exec 环境参数注入的代理**

代理策略目前只检查 `req.Command`，但 `exec_command` 同时允许传入 `env`。因此，使用 `env: {"HTTPS_PROXY":"http://proxy:8080"}` 配合普通 `curl` 命令，就能通过临时代理发出请求，而不会触发 `blocksNetworkProxy`。

宿主实现随后会把这些显式参数合并进既有环境，建议在合并前检查工具调用传入的覆盖值，或在 `CommandRequest` 中保留其来源。应继续允许运行环境预先配置的代理，但拒绝调用参数中新设置的非空 `HTTP_PROXY`、`HTTPS_PROXY` 和 `ALL_PROXY`，并覆盖宿主与沙箱两条执行路径。

</details>

File: openclaw/internal/octool/policy.go [hunk 8]
```diff
@@ -118,6 +187,437 @@ func blocksSystemPackageInstall(command string) bool {
 	return blocksSystemPackageInstallDepth(normalizePolicyCommand(command), 0)
 }
 
+func blocksSearchResultHTTP(command string) (string, bool) {
+	command = normalizePolicyCommand(command)
+	if command == "" || !usesCommandLineHTTPClient(command, 0) {
+		return "", false
+	}
+	for _, raw := range httpURLPattern.FindAllString(command, -1) {
```
Comment (Rememorio) [inline openclaw/internal/octool/policy.go:196]: **P1: Cover search queries constructed outside the literal URL**

This guard requires the literal URL to contain the provider's query key. A normal command such as `curl -G --data-urlencode q=openclaw https://www.google.com/search` sends the same Google result-page request, but `searchresult.Match` sees no `q` in the scanned URL and allows it.

Handle common client query-building flags, or treat exact known result endpoints as blocked when the HTTP client supplies query/form data separately. Add a regression for curl's `-G` plus `--data-urlencode` form so the policy covers ordinary command construction rather than only inline query strings.

<details><summary>中文</summary>

**覆盖未直接写入 URL 的搜索查询**

当前防护要求 URL 字面量中直接包含搜索参数。像 `curl -G --data-urlencode q=openclaw https://www.google.com/search` 这样的常规写法会发出同样的 Google 搜索结果请求，但 `searchresult.Match` 扫描到的 URL 中没有 `q`，因此会被放行。

建议识别常见 HTTP 客户端在 URL 外构造查询参数的选项，或者在客户端另外提供查询/表单数据时，将已知的精确搜索端点直接视为受限路由。至少增加 `curl -G` 与 `--data-urlencode` 组合的回归测试。

</details>

File: tool/webfetch/httpfetch/fetch.go [hunk 9]
```diff
@@ -389,6 +457,248 @@ func unsupportedContentTypeError(contentType string) string {
...
+	if containsAny(
+		lower,
+		"verify you are human",
+		"checking if the site connection is secure",
+		"review the security of your connection",
+		"enable javascript and cookies to continue",
+	) {
+		return "human-verification challenge", true
+	}
+	if containsAny(
```
Comment (Rememorio) [inline tool/webfetch/httpfetch/fetch.go]: **P1: Avoid treating any anti-bot mention as a blocked page**

This predicate classifies any response containing `anti-bot`, `bot check`, or similar text as a challenge, without requiring page-shape or verification cues. With OpenClaw's default `detect_blocked_pages: true`, a legitimate security article or JSON API response discussing anti-bot systems becomes an error; the new blocked-route callback then records that host and skips its other URLs for the rest of the invocation.

Require stronger co-signals or constrain these weak phrases to short challenge-like responses. Add an end-to-end OpenClaw test where a successful article mentions anti-bot/CAPTCHA terminology and verify both that its content is returned and that a subsequent URL on the same host is still fetched.

<details><summary>中文</summary>

**避免将任意反机器人术语误判为挑战页**

这里在没有页面结构或验证提示等辅助信号的情况下，只要响应中出现 `anti-bot`、`bot check` 等字样就会判定为挑战页。在 OpenClaw 默认开启 `detect_blocked_pages` 时，正常的安全文章或讨论反机器人机制的 JSON API 响应会被转成错误；随后新增的路由记忆回调还会记录该主机，导致本次调用后续对同一主机的其他 URL 也被跳过。

建议要求更强的组合信号，或仅在短小、明显类似挑战页的响应中使用这些弱特征。应增加 OpenClaw 端到端测试：成功页面包含反机器人或 CAPTCHA 术语时仍返回正文，并且同一主机的下一个 URL 仍会实际抓取。

</details>

File: openclaw/internal/browser/tool.go [hunk 10]
```diff
@@ -2903,13 +2945,90 @@ func (t *Tool) textResult(
 	)
 	result.Untrusted = true
 	result.Warning = untrustedBrowserWarning
-	result.Text = wrapUntrustedText(
-		extractText(raw),
-		intValue(maxChars),
-	)
+	text := extractText(raw)
+	if reason, ok := blockedBrowserPageReason(text); ok {
```
Comment (Rememorio) [inline openclaw/internal/browser/tool.go]: **P2: Make browser challenge rewriting configurable and action-aware**

Previously browser text results were returned with the untrusted-content wrapper. They are now unconditionally rewritten to `state=blocked` whenever this generic matcher fires, with no browser equivalent of web-fetch's `detect_blocked_pages` opt-out. Because every `textResult` passes through here, even console, cookie, storage, or download output mentioning `anti-bot` can be mislabeled as a blocked page.

Gate this behavior behind an explicit browser configuration option and apply it only to actions whose result represents page content or navigation state. Add compatibility coverage for the disabled setting and a non-page action containing challenge terminology.

<details><summary>中文</summary>

**让浏览器挑战页改写可配置且区分操作类型**

此前浏览器文本结果只会加上不可信内容提示；现在只要通用匹配器命中，就会无条件改写为 `state=blocked`，而浏览器配置中没有类似 web-fetch 的 `detect_blocked_pages` 关闭选项。由于所有 `textResult` 都经过这里，控制台、Cookie、存储或下载结果中只要提到 `anti-bot`，也可能被错误标记为页面受阻。

建议增加显式的浏览器配置开关，并仅对确实代表页面内容或导航状态的操作启用该检测。需要补充关闭开关后的兼容性测试，以及非页面操作返回挑战页相关术语时不被误判的测试。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 5 inline comments below. The comments focus on issues worth addressing before merge.

Reviewed the new browser/web-fetch route guards, blocked-route memory, DuckDuckGo filtering, and exec-command policy. The main risks are enforcement bypasses through alternate navigation/query construction, proxy settings supplied through tool environment arguments, and challenge detection that can poison legitimate same-host fetches. Add targeted regressions for redirects/click navigation, curl query flags, explicit proxy env overrides, and ordinary security-related content.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 5 条行内评论。评论聚焦在合并前值得处理的问题。

已审查浏览器与 web_fetch 路由防护、阻断路由记忆、DuckDuckGo 结果过滤及命令执行策略。主要风险是替代导航或查询构造方式可绕过防护、工具参数中的代理环境变量未受限制，以及挑战页误判会连带阻断同一主机的正常内容。建议补充重定向与点击导航、curl 查询参数、显式代理环境变量，以及普通安全类文章的回归测试。

</details>

File: openclaw/internal/browser/tool.go (hunk 6 above)
Comment (WineChord) [reply openclaw/internal/browser/tool.go:1631]: fixed in 79b32086b: browser-server and extension results now expose the resulting URL, and navigate/open/act re-check search-page policy after redirects or page-changing actions; redirect and click regressions added

File: openclaw/internal/octool/policy.go (hunk 7 above)
Comment (WineChord) [reply openclaw/internal/octool/policy.go]: fixed in 79b32086b: explicit non-empty HTTP_PROXY, HTTPS_PROXY, and ALL_PROXY request env overrides are rejected case-insensitively; host and sandbox tool coverage added

File: openclaw/internal/octool/policy.go (hunk 8 above)
Comment (WineChord) [reply openclaw/internal/octool/policy.go:196]: fixed in 79b32086b: curl query-building data flags are recognized for exact search endpoints, including -G --data-urlencode q= regression coverage

File: tool/webfetch/httpfetch/fetch.go (hunk 9 above)
Comment (WineChord) [reply tool/webfetch/httpfetch/fetch.go]: fixed in 79b32086b: weak anti-bot phrases now require a short response plus challenge/block/verification co-signals; security-article coverage preserves normal fetch content

File: openclaw/internal/browser/tool.go (hunk 10 above)
Comment (WineChord) [reply openclaw/internal/browser/tool.go]: fixed in 79b32086b: browser detect_blocked_pages is explicit and defaults on, can be disabled, and challenge rewriting is limited to page-content actions; console compatibility coverage added

**Review by Flash-LHR — APPROVED** [review]
LGTM
